docs/framework/how-to/file-uploads.md

```diff
@@ -0,0 +1,195 @@
+---
+title: File Uploads
+---
+
+# File Uploads
+
+Handle file uploads in your React Router applications. This guide uses some packages from the [Remix The Web][remix-the-web] project to make file uploads easier.
+
+_Thank you to David Adams for [his original guide](https://programmingarehard.com/2024/09/06/remix-file-uploads-updated.html/) on how to implement file uploads in Remix. You can refer to it for even more examples._
+
+## Basic File Upload
+
+👉 **Add the form data parser**
+
+`form-data-parser` is a wrapper around `request.formData()` that provides streaming support for handling file uploads.
+
+```shellscript
+npm i @mjackson/form-data-parser
+```
+
+[See the `form-data-parser` docs for more information][form-data-parser]
+
+👉 **Create a route with an upload action**
+
+The `parseFormData` function takes an `uploadHandler` function as an argument. This function will be called for each file upload in the form.
+
+<docs-warning>
+
+You must set the form's `enctype` to `multipart/form-data` for file uploads to work.
+
+</docs-warning>
+
+```tsx filename=routes/user.$id.tsx
+import {
+  type FileUpload,
+  parseFormData,
+} from "@mjackson/form-data-parser";
+
+export async function action({
+  request,
+}: ActionFunctionArgs) {
+  const uploadHandler = async (fileUpload: FileUpload) => {
+    if (fileUpload.fieldName === "avatar") {
+      // process the upload and return a File
+    }
+  };
+
+  const formData = await parseFormData(
+    request,
+    uploadHandler
+  );
+  // 'avatar' has already been processed at this point
+  const file = formData.get("avatar");
+}
+
+export default function Component() {
+  return (
+    <Form method="post" encType="multipart/form-data">
+      <input type="file" name="avatar" />
+      <button>Submit</button>
+    </Form>
+  );
+}
+```
+
+## Local Storage Implementation
+
+👉 **Add the storage package**
+
+`file-storage` is a key/value interface for storing [File objects][file] in JavaScript. Similar to how `localStorage` allows you to store key/value pairs of strings in the browser, file-storage allows you to store key/value pairs of files on the server.
+
+```shellscript
+npm i @mjackson/file-storage
+```
+
+[See the `file-storage` docs for more information][file-storage]
+
+👉 **Create a storage configuration**
+
+Create a file that exports a `LocalFileStorage` instance to be used by different routes.
+
+```ts filename=file-storage.server.ts
```

Comment:
Might wanna call this file `avatar-storage.server.ts` or something similar. It's a bit more descriptive than just "file storage".

Reply:
Good idea, updated!